state/indexer/indexer_service.go

```diff
@@ -51,43 +54,47 @@ func (is *Service) OnStart() error {
 
 	go func() {
 		for {
-			msg := <-blockHeadersSub.Out()
-
-			eventDataHeader := msg.Data().(types.EventDataNewBlockHeader)
-			height := eventDataHeader.Header.Height
-			batch := NewBatch(eventDataHeader.NumTxs)
-
-			for i := int64(0); i < eventDataHeader.NumTxs; i++ {
-				msg2 := <-txsSub.Out()
-				txResult := msg2.Data().(types.EventDataTx).TxResult
-
-				if err = batch.Add(&txResult); err != nil {
-					is.Logger.Error(
-						"failed to add tx to batch",
-						"height", height,
-						"index", txResult.Index,
-						"err", err,
-					)
+			select {
+			case <-is.doneChan:
```

Comment:
This issue was caught by the leak checker. Without it, the goroutine started on line 55 would never return. It would block forever on the `blockHeaderSub.Out` channel.

Reply:
That seems right, but the way we are managing concurrency in this type is very fragile. Ideally correct maintenance of concurrent state should not rely on the user of the type calling another method on the interface. This isn't like calling wait to reap the status of a finished task: Here failing to invoke the stop hook will actually deadlock, not just orphan some resources. 

Can we reorganize this so that the concurrency is either (a) not necessary (e.g., supplied by the caller) or (b) managed fully within the scope where the goroutine is spawned? I would like to make a more concrete suggestion, but right now I don't really get what is going on here.

Reply:
So, I think the issue here pretty directly relates to the assumptions that currently exist around the `service.Service` interface. This interface has an `OnStart` method and a `Stop` and `OnStop` method. The `OnStart` method appears to not generally be called in a groutine when it is invoked. In general these calls appear to return immediately. Here's an example of such an [OnStart in the wal](https://github.com/tendermint/tendermint/blob/69239c55c95d50655ece4b5a25d024b88549c751/internal/consensus/wal.go#L133). Types that implement service rely on the caller being responsible for shutting down the `service` after some time. The service starts the concurrency, but the caller is responsible for it. I agree that this pattern is a bit inverted but I'm not sure it's really worth completely untying in this pull request.